plonk-core/src/proof_system/pi.rs

```diff
@@ -114,6 +114,34 @@ where
         let evals = self.as_evals(n);
         DensePolynomial::from_coefficients_vec(domain.ifft(&evals))
     }
+
+    /// Constructs [`PublicInputs`] from a positions and a values.
+    ///
+    /// Panics if the positions and values have different lenghts or if
+    /// several values try to be inserted in the same position.
+    pub fn from_val_pos<T>(pos: &[usize], vals: &[T]) -> Result<Self, Error>
+    where
+        T: ToConstraintField<F>,
+    {
+        let mut pi = Self::new();
+        pos.iter().zip_eq(vals).try_for_each(
+            |(p, v)| -> Result<(), Error> {
+                pi.add_input(*p, v)?;
+                Ok(())
+            },
+        )?;
+        Ok(pi)
+    }
+
+    /// Returns the position of non-zero PI values.
+    pub fn get_pos(&self) -> Vec<usize> {
+        self.values.keys().cloned().collect()
+    }
+
+    /// Returns the non-zero PI values.
+    pub fn get_vals(&self) -> Vec<F> {
+        self.values.values().cloned().collect()
+    }
```

Comment:
Try to return references to the `Vec`s and then it's up to the consumer to clone or not.
Otherwise, the functions can return: `& impl Iterator<Item=F>`. So that you can call `.iter()` to the returned result and clone if necessary.

Reply:
Good point!